### DISCUSSION TITLE
`iterator_policies::sequence_fast_readonly` not default constructible?

### DISCUSSION THREAD

**Discussion by zwhfly** [pybind/pybind11#3850] [Ideas]
I found out that `pybind11::detail::iterator_policies::sequence_fast_readonly` is not default constructible.
This makes the type returned by `pybind11::args::begin()` not satisfying the `std::input_or_output_iterator` concept.

I added the default ctor directly in the header file: https://github.com/zwhfly/pybind11/commit/bfa326a8851bf37c0c7b1825f1612d1153326fdf, and made a test.

Given `using Iter = decltype(std::declval<pybind11::args>().begin());`, this change makes `std::constructible_from<Iter>` true, which makes `std::default_initializable<Iter>` true, which makes `std::weakly_incrementable<Iter>` true, which makes `std::input_or_output_iterator<Iter>` true, which makes `pybind11::args` satisfy `std::ranges::range` concept, which makes the following code work:

```c++
void f(pb11::args args)
{
    namespace views = std::ranges::views;
    auto vu = args | views::transform(
        [](pybind11::handle const & h) -> MyClass const & { return h.cast<MyClass const &>(); });
    some_library_function(vu.begin(), vu.end());
}
```

I'm not familiar with the source code of pybind11, so I wonder would this change cause any bad effect that I don't know of?
Could it be added?

**Comment by Skylion007** [comment]
@zwhfly Feel free to add. It seems like the use case could be compelling, just please be sure to add a test in your PR as well.

**Reply by Skylion007** [reply]
@zwhfly It actually appears as though the C++20 standard has been updated so that this is no longer an issue. See: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2325r3.html
